File: ITEM 40-49/ITEM45 OpenCV_ImgHash/OpenCV_ImgHash.py
# -*- coding: utf-8 -*-
"""
Created on Thu Aug 27 19:03:21 2020

@author: luohenyueji
"""

import cv2


def test_one(title, a, b):
    # 创建类
    if "AverageHash" == title:
        hashFun = cv2.img_hash.AverageHash_create()
    elif "PHash" == title:
        hashFun = cv2.img_hash.PHash_create()
    elif "MarrHildrethHash" == title:
        hashFun = cv2.img_hash.MarrHildrethHash_create()
    elif "RadialVarianceHash" == title:
        hashFun = cv2.img_hash.RadialVarianceHash_create()
    elif "BlockMeanHash" == title:
        hashFun = cv2.img_hash.BlockMeanHash_create()
    elif "ColorMomentHash" == title:
        hashFun = cv2.img_hash.ColorMomentHash_create()

    tick = cv2.TickMeter()
    print("=== " + title + " ===")

    tick.reset()
    tick.start()
    # # 计算图a的哈希值
    hashA = hashFun.compute(a)
    tick.stop()
    print("compute1: " + str(tick.getTimeMilli()) + " ms")

    tick.reset()
    tick.start()
    # 计算图b的哈希值
    hashB = hashFun.compute(b)
    tick.stop()
    print("compute2: " + str(tick.getTimeMilli()) + " ms")
    # 比较两张图像哈希值的距离
    print("compare: " + str(hashFun.compare(hashA, hashB)))


def main():
    inputImg = cv2.imread("./image/img1.jpg")
    targetImg = cv2.imread("./image/img4.jpg")

    if inputImg is None or targetImg is None:
        print("check input image")
        return

    test_one("AverageHash", inputImg, targetImg)
    test_one("PHash", inputImg, targetImg)
    test_one("MarrHildrethHash", inputImg, targetImg)
    test_one("RadialVarianceHash", inputImg, targetImg)
    test_one("BlockMeanHash", inputImg, targetImg)
    test_one("ColorMomentHash", inputImg, targetImg)


if __name__ == '__main__':
    main()

File: ITEM 40-49/ITEM45 OpenCV_ImgHash/OpenCV_ImgHash.cpp
#include <opencv2/opencv.hpp>
#include <opencv2/img_hash.hpp>

#include <iostream>

using namespace cv;
using namespace cv::img_hash;
using namespace std;

template <typename T>
inline void test_one(const std::string &title, const Mat &a, const Mat &b)
{
	cout << "=== " << title << " ===" << endl;
	TickMeter tick;
	Mat hashA, hashB;
	// 模板方便重复利用
	Ptr<ImgHashBase> func;
	func = T::create();

	tick.reset();
	tick.start();
	// 计算图a的哈希值
	func->compute(a, hashA);
	tick.stop();
	cout << "compute1: " << tick.getTimeMilli() << " ms" << endl;

	tick.reset();
	tick.start();
	// 计算图b的哈希值
	func->compute(b, hashB);
	tick.stop();
	cout << "compute2: " << tick.getTimeMilli() << " ms" << endl;

	// 比较两张图像哈希值的距离
	cout << "compare: " << func->compare(hashA, hashB) << endl << endl;
}

int main()
{
	// 打开两张图像进行相似度比较
	Mat input = imread("./image/img1.jpg");
	Mat target = imread("./image/img4.jpg");

	// 通过不同方法比较图像相似性
	test_one<AverageHash>("AverageHash", input, target);
	test_one<PHash>("PHash", input, target);
	test_one<MarrHildrethHash>("MarrHildrethHash", input, target);
	test_one<RadialVarianceHash>("RadialVarianceHash", input, target);
	test_one<BlockMeanHash>("BlockMeanHash", input, target);
	test_one<ColorMomentHash>("ColorMomentHash", input, target);

	system("pause");
	return 0;
}